### PR TITLE
convert CrossValidation PP to use new data objects.

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
convert the cross validation pp to use the new data objects
##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

**Comment by wangcj05** [comment]
Tests Results:
```
**** THE FOLLOWING TESTS HAVE BEEN REMOVED IN 2ND FILE ****
**** Test framework/ensembleModelTests.testEnsembleModelFullyCorrVars | has been REMOVED! It is present in 1st file BUT NOT in 2nd file !!!!
*****************
Number of success first file:
218
Number of success second file:
290
NUMBER OF SUCCESSES INCREASED: +72
**** NUMBER OF TESTS IN THE TWO FILES ARE DIFFERENT: ****
**** first file :533 /= second file: 532****
```

**Comment by wangcj05** [comment]
The missing test is the one added recently with PR #465. Should be fixed when rebase.

File: framework/PostProcessors/CrossValidation.py [hunk 1]
```diff
@@ -185,6 +187,7 @@ def inputToInternal(self, currentInp, full = False):
       understandable by this pp.
       @ In, currentInp, list or DataObject, data object or a list of data objects
       @ In, full, bool, optional, True to retrieve the whole input or False to get the last element of the input
+        TODO, full should be removed
```
Comment (PaulTalbot-INL) [inline framework/PostProcessors/CrossValidation.py]: Is this a dataobject rework TODO, or a future TODO for devel (in which case we might need an issue for it)

**Comment by wangcj05** [comment]
Here are the full test results:
```
**** SUMMARY ****
*********************************
**** Unmatched tests are:

framework/ensembleModelTests.testEnsembleModelFullyCorrVars
*********************************
**** THE FOLLOWING TESTS CHANGED OUTCOMES ****
**** Test framework/PostProcessors/CrossValidations.labelKFold | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/PostProcessors/CrossValidations.labelShuffleSplit | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/PostProcessors/CrossValidations.leaveOneLabelOut | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/PostProcessors/CrossValidations.leaveOneOut | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/PostProcessors/CrossValidations.leavePLabelOut | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/PostProcessors/CrossValidations.leavePOut | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/PostProcessors/CrossValidations.shuffleSplit | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/PostProcessors/CrossValidations.simpleCrossValidation | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/PostProcessors/CrossValidations.stratifiedKFold | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/PostProcessors/CrossValidations.stratifiedShuffleSplit | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM.timeDepRom | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.Epanechnikov | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.Gaussian | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.SVM | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.Silverman | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.biweight | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.cosine | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.exponential | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.logistic | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.parallel | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothBiweight | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothCosine | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothEpanechnikov | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothExponential | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothGaussian | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothLogistic | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothSVM | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothSilverman | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothTriangular | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothTricube | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothTriweight | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.smoothUniform | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.triangular | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.tricube | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.triweight | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/MSR.uniform | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.BernoulliNB | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.DTC | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.ETC | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.GP | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.GaussianNB | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.KNC | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.KNR | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearARD | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearBayesianRidge | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearElasticNet | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearElasticNetCV | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearLARS | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearLARSCV | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearLasso | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearLassoCV | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearLassoLARS | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearLassoLARSCV | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearLassoLARSIC | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearLogisticRegression | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearMultiTaskElasticNet | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearMultiTaskLasso | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearOMP | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearOMPCV | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearPAC | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearPAR | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearPerceptron | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearRegression | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearRidgeC | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearRidgeCCV | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearRidgeR | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearRidgeRCV | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearSGDC | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearSGDR | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/SKLearn.linearSVC | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/Sobol.staticSobolRomSmolyak | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/ROM/Sobol.staticSobolRomTensor | has changed outcome from <FAILED> to <OK> !!!!
*****************
**** THE FOLLOWING TESTS HAVE BEEN REMOVED IN 2ND FILE ****
**** Test framework/ensembleModelTests.testEnsembleModelFullyCorrVars | has been REMOVED! It is present in 1st file BUT NOT in 2nd file !!!!
*****************
Number of success first file:
218
Number of success second file:
290
NUMBER OF SUCCESSES INCREASED: +72
**** NUMBER OF TESTS IN THE TWO FILES ARE DIFFERENT: ****
**** first file :533 /= second file: 532****
```

**Comment by wangcj05** [comment]
@PaulTalbot-INL @alfoa @mandd Anyone wants to take a look at this PR?

File: framework/PostProcessors/CrossValidation.py [hunk 2]
```diff
@@ -358,89 +342,7 @@ def collectOutput(self,finishedJob, output):
...
-      output.open('w')
-      if outputExtension == 'xml':
-        self._writeXML(output, outputDict)
-      else:
-        separator = ' ' if outputExtension != 'csv' else ','
-        self._writeText(output, outputDict, separator)
-    else:
-      self.raiseAnError(IOError, 'Output type ', str(output.type), ' can not be used for postprocessor', self.name)
+    print("Debug Realization: ")
+    print(outputDict)
```
Comment (PaulTalbot-INL) [inline framework/PostProcessors/CrossValidation.py]: Debug statements should be removed.

File: framework/PostProcessors/CrossValidation.py [hunk 3]
```diff
@@ -306,45 +295,40 @@ def run(self, inputIn):
...
+            self.raiseAnError(IOError, "The metric: ", metricInstance.name, " can not be used, the accepted metric types are: ", str(self.validMetrics))
+
+        varName = 'cv' + '_' + metricInstance.name + '_' + cvEstimator.name
+        if varName not in outputDict.keys():
+          outputDict[varName] = []
+        outputDict[varName].append(metricValue[0])
+
     scoreDict = {}
-    if not self.cvScores:
-      return outputDict
```
Comment (alfoa) [inline framework/PostProcessors/CrossValidation.py]: So you removed the possibility to get the scores for each fold?

Comment (alfoa) [reply framework/PostProcessors/CrossValidation.py]: Can you put the "update documenation" in the todolist for this work? (The shared sheet in google doc)

File: framework/PostProcessors/CrossValidation.py (hunk 1 above)
Comment (wangcj05) [reply framework/PostProcessors/CrossValidation.py]: I think this is a data object rework TODO.

File: framework/PostProcessors/CrossValidation.py (hunk 3 above)
Comment (wangcj05) [reply framework/PostProcessors/CrossValidation.py]: Yes, I think this PP should return a single quantity to tell the user if the ROM is good or not.

File: framework/PostProcessors/CrossValidation.py (hunk 1 above)
Comment (PaulTalbot-INL) [reply framework/PostProcessors/CrossValidation.py]: Okay.  What is the change need before it's acted on?  Is this a "once someone has time and needs speed" changes?

File: framework/PostProcessors/CrossValidation.py (hunk 3 above)
Comment (wangcj05) [reply framework/PostProcessors/CrossValidation.py]: I have updated the shared excel sheet.

File: framework/PostProcessors/CrossValidation.py (hunk 2 above)
Comment (wangcj05) [reply framework/PostProcessors/CrossValidation.py]: removed.

File: framework/PostProcessors/CrossValidation.py (hunk 1 above)
Comment (wangcj05) [reply framework/PostProcessors/CrossValidation.py]: When we start to finalize the inputToInternal, and remove the collectOutput. The final clean up I think.

File: framework/PostProcessors/CrossValidation.py (hunk 3 above)
Comment (alfoa) [reply framework/PostProcessors/CrossValidation.py]: I agree it should by default, but honestly I can use the cross validation (k-fold) also as metric to understand how homogeneously my regression is able to capture the underlying model. I would have changed the default (to be the single scalar) but not remove the possibility to get the stats from all the folds...

**Comment by wangcj05** [comment]
@alfoa As your result, I have changed the CV pp to output results for each fold, please let me know if you have additional comments.

**Pull request merged**